walk-in-interview/src/main/webapp/account/firebase-auth.js

```diff
@@ -92,56 +80,95 @@ Auth.addPhoneAuthUI = (elementId, successPath) => {
       signInSuccessWithAuthResult: (authResult, redirectUrl) => {
         // TODO(issue/89): add new user pages for name/skills
         // for now, this will only redirect to homepage for exisiting users
-        return !authResult.additionalUserInfo.isNewUser;
+
+        // TODO(issue/100): set the cookie at the server side instead
+        setCookie(USER_TYPE_COOKIE_PARAM, TYPE_APPLICANT);
+
+        if (authResult.additionalUserInfo.isNewUser) {
+          // Informs user
+          alert(newUserInfo);
+        }
+
+        return true;
       },
+
+      signInFailure: (error) => {
+        console.error(error.message);
+
+        alert(STRINGS['sign-in-failure']);
+      }
     },
     signInSuccessUrl: successPath,
   });
 };
 
-/**
- * This will create a new applicant account.
- *
- * @param {String} phoneNumber The phone number for the new applicant account.
- * @param {Object} appVerifier The recaptcha verifier.
- */
-Auth.createApplicantAccount = (phoneNumber, appVerifier) => {
-  // TODO(issue/79): set up phone number sign in with otp and recaptcha
-};
-
-/**
- * This will sign into an existing applicant account.
- *
- * @param {String} phoneNumber The number for the existing applicant account.
- * @param {Object} appVerifier The recaptcha verifier.
- */
-Auth.signIntoApplicantAccount = (phoneNumber, appVerifier) => {
-  // TODO(issue/79): set up phone number sign in with otp and recaptcha
-};
-
 /**
  * Signs out the current user.
- *
- * @param {String} elementId The div in which to show the signed out status.
  */
-Auth.signOutCurrentUser = (elementId) => {
+Auth.signOutCurrentUser = () => {
   firebase.auth().signOut().then(() => {
     console.log('sign out successful');
-    document.getElementById(elementId).innerText = STRINGS['sign-out-success'];
+    // TODO(issue/100): set the cookie at the server side instead
+    setCookie(USER_TYPE_COOKIE_PARAM, TYPE_NO_USER);
+    alert(STRINGS['sign-out-success']);
   }).catch((error) => {
     console.error(error);
   });
+  Auth.subscribeToUserAuthenticationChanges();
 };
 
 /**
  * Checks the user sign in status.
+ * Makes a POST request to create a session cookie when the user signs in.
+ * Makes a POST request to clear the session cookie when the user signs out.
  */
-Auth.checkCurrentUser = () => {
+Auth.subscribeToUserAuthenticationChanges = () => {
+  // To make sure it only triggered once when sign in and out
+  var authFlag = true;
+
   firebase.auth().onAuthStateChanged((firebaseUser) => {
-    if (firebaseUser) {
-      console.log('signed in', firebaseUser);
-    } else {
-      console.log('not signed in');
+    if (authFlag) {
```

Comment:
this logic doesn't look right. What about:
```
var currentUser =  null;
var currentCsrfToken = null;

subscribe(firebaseUser -> {
     // logged out.
     if( firebaseUser == null ){
        if( currentUser == null ){
             // already logged out, no-op
             return;
        }else{
        	currentUser = null;
        	Auth.postIdTokenToSessionLogout();
        }
	    return;
    }

    // logged in
    currentUser = firebaseUser;
    firebaseUser.getIdToken(idToken -> {

         csrfToken = getCookie();
         if( currentCsrfToken == csrfToken ){
              return; // same login as before, no need to create a new session
         }
         currentCsrfToken = csrfToken;
         [...]
         Auth.postIdTokenToSessionLogin();
    })

}

```

Reply:
When i was experimenting the code, I figured out that within a short period of time, i.e. before a ID token expires, 2 different accounts also share the same id token & session cookie (since the session cookie is generated via id token, which is the same), but the uid of account is different(i.e. even if the id token is the same, can still locate the correct account respectively). Hence i think only need to check the log in status and create/clear session cookie accordingly, do not need to check if id token is the same